vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/ConnectionFactoryBase.java

```diff
@@ -40,8 +41,7 @@ public abstract class ConnectionFactoryBase implements ConnectionFactory {
   protected final Map<String, String> properties;
   protected final SqlConnectOptions options;
   protected final SocketAddress server;
-  protected final String user;
-  protected final String password;
+  protected final SqlCredentialsProvider credentialsProvider;
```

Comment:
It don't believe we should have this in the options class.

If you create a connection manually, the credentials provider is redundant: you could get the latest credentials before creating the connect options.

So the crendentials provider really makes sense for pools. Then it should be set the same way as the connection provider, i.e:

```java
  @Fluent
  Pool connectionProvider(Function<Context, Future<SqlConnection>> provider);
```

This way when RxJava / Mutiny users customize their pool , will get a method that works well with their codebase.

Reply:
It's directly affecting the SqlConnectionOptions and is an alternative to the username/password options. I believe this is the best place for it.

> If you create a connection manually, the credentials provider is redundant: you could get the latest credentials before creating the connect options.

Yes you could. It's also pretty convenient for people using external credentials systems, that might also open a direct connection (e.g. for migration or notifications), to share a single configured `SqlConnectOptions`. Additionally, this colocates all credentials related parameters into a single options object.

> This way when RxJava / Mutiny users customize their pool , will get a method that works well with their codebase.
This already works quite well with those codebases. I've had this wired all the way through in Quarkus since I opened the PR and it's tidy and convenient. The only downside is, as you pointed out, there is currently no codegen support; but that will also requires a bump to the related smallrye mutiny library.